core/src/main/scala/scalaz/Semigroup.scala

```diff
@@ -22,6 +22,33 @@ trait Semigroup[F]  { self =>
   def append(f1: F, f2: => F): F
 
   // derived functions
+
+  /**
+   * For `n = 0`, `value`
+   * For `n = 1`, `append(value, value)`
+   * For `n = 2`, `append(append(value, value), value)`
+   *
+   * The default definition uses peasant multiplication, exploiting associativity to only
+   * require `O(log n)` uses of [[append]]
+   */
+  def multiply1(value: F, n: Int): F = {
+    @scala.annotation.tailrec
+    def f(x: F, y: Int): F = y match {
+      case y if (y % 2 == 0) => f(append(x, x), y / 2)
```

Comment:
As this has an impact on readability, I'd like another opinion on those optimizations before I make the changes... someone?